lib/smart-contract-client.js

```diff
@@ -0,0 +1,44 @@
+import { MINER_TO_PEERID_CONTRACT_ADDRESS, RPC_AUTH, RPC_URL } from './constants.js'
+import assert from 'assert'
+import { ethers } from 'ethers'
+
+// ABI for the MinerPeerIDMapping contract (minimal ABI with just the method we need)
+// Docs for smart contract: https://github.com/filecoin-project/curio/blob/395bc47d0f585cbc869fd4671dc05b1b2f4b18c2/market/ipni/spark/sol/README.md
+// Reasoning for smart contract: https://docs.curiostorage.org/curio-market/ipni-interplanetary-network-indexer-provider#ipni-provider-identification
+const abi = [
+  'function getPeerData(uint64 minerID) view returns (tuple(string peerID, bytes signature))',
+]
+
+// Create a custom JsonRpcProvider with authorization header
+const fetchRequest = new ethers.FetchRequest(RPC_URL)
+fetchRequest.setHeader('Authorization', `Bearer ${RPC_AUTH}`)
+const provider = new ethers.JsonRpcProvider(fetchRequest)
+
+// Create contract instance
+const defaultClient = new ethers.Contract(MINER_TO_PEERID_CONTRACT_ADDRESS, abi, provider)
+
+/**
+ * Query the smart contract for the peer ID mapping
+ * @param {string} minerID - The miner ID (as string, will be converted to uint64)
+ * @param {object} [options]
+ * @param {ethers.Contract} [options.smartContract] - A smart contract client to use instead of the default one
+ * @returns {Promise<string|null>} The peer ID from the contract or empty string if not found
+ */
+export async function getIndexProviderPeerIdFromSmartContract(minerID, { smartContract } = {}) {
+  try {
+    const contractClient = smartContract ?? defaultClient
+    assert(contractClient, 'smartContract must be initialized')
+    // Convert minerID string (like 'f01234') to numeric ID
+    const numericID = parseInt(minerID.replace('f0', ''))
+    assert(!isNaN(numericID), `minerID must be "f0{number}". Actual value: "${minerID}"`)
+
+    // Call the contract function
+    const peerData = await contractClient.getPeerData(numericID)
+    // TODO: Check if peerData.signature is valid
+    return peerData.peerID ?? null
```

Comment:
See https://github.com/filecoin-project/curio/blob/395bc47d0f585cbc869fd4671dc05b1b2f4b18c2/market/ipni/spark/sol/contracts/MinerPeerIDMapping.sol#L89-L92. If the peer data can't be found, `peerID` will be returned as default value = empty string.

```suggestion
    return peerData.peerID !== '' ? peerData.peerID : null
```

Reply:
Yes, and if the returned value is an empty string then we want to return an empty string, no? 
An empty string is different to null and undefined, meaning if we convert an empty string to null we loose information about the returned data.

Reply:
Prior discussion:

https://github.com/CheckerNetwork/spark-checker/pull/117#discussion_r1989554294

Reply:
If we want to return the empty string, then we should remove this as dead code

```suggestion
    return peerData.peerID
```